src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,285 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.IntroduceUsingStatement
+{
+    internal abstract class AbstractIntroduceUsingStatementCodeRefactoringProvider<TStatementSyntax, TLocalDeclarationSyntax> : CodeRefactoringProvider
+        where TStatementSyntax : SyntaxNode
+        where TLocalDeclarationSyntax : TStatementSyntax
+    {
+        protected abstract bool IsBlockLike(SyntaxNode node);
+        protected abstract SyntaxList<TStatementSyntax> GetStatements(SyntaxNode blockLike);
+        protected abstract SyntaxNode WithStatements(SyntaxNode blockLike, SyntaxList<TStatementSyntax> statements);
+
+        protected abstract TStatementSyntax CreateUsingStatement(TLocalDeclarationSyntax declarationStatement, SyntaxTriviaList sameLineTrivia, SyntaxList<TStatementSyntax> statementsToSurround);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(document, span, context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Introduce_using_statement,
+                    cancellationToken => IntroduceUsingStatementAsync(document, span, cancellationToken)));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax =
+                root.FindNode(selection)?.GetAncestor<TLocalDeclarationSyntax>()
+                ?? root.FindTokenOnLeftOfPosition(selection.End).GetAncestor<TLocalDeclarationSyntax>();
+
+            if (declarationSyntax is null)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var disposableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType is null)
+            {
+                return default;
+            }
```

Comment:
i believe should not be possible.  'special types' ether come back, or come back as error types.  but i'm ok being paranoid here.